contracts/leaser/src/migrate.rs

```diff
@@ -2,66 +2,97 @@ use lease::api::MigrateMsg;
 use platform::batch::Batch;
 use sdk::cosmwasm_std::Addr;
 
-use crate::{error::ContractError, result::ContractResult};
+use crate::{error::ContractError, msg::NbInstances, result::ContractResult};
 
-pub fn migrate_leases<I>(leases: I, lease_code_id: u64) -> ContractResult<Batch>
+#[derive(Default)]
+#[cfg_attr(test, derive(Debug, Eq, PartialEq))]
+pub struct MigrationResult {
+    pub msgs: Batch,
+    /// None if the number of processed instances is less than the `max_leases`
+    pub last_instance: Option<Addr>,
+}
+
+pub fn migrate_leases<I>(
+    leases: I,
+    lease_code_id: u64,
+    max_leases: NbInstances,
+) -> ContractResult<MigrationResult>
 where
     I: Iterator<Item = ContractResult<Addr>>,
 {
-    let no_msgs = MigrateBatch::new(lease_code_id);
-    let migrated_msgs = leases.fold(no_msgs, MigrateBatch::migrate_lease);
+    let no_msgs = MigrateBatch::new(lease_code_id, max_leases);
+    let migrated_msgs = leases
+        .take(max_leases.try_into()?)
+        .fold(no_msgs, MigrateBatch::migrate_lease);
```

Comment:
`Iterator::try_fold` is stable, `Iterator::try_reduce` isn't, in case that was the reason to avoid it.